templates/serve-stable-diffusion-aica/main.py

```diff
@@ -0,0 +1,86 @@
+import logging
+from io import BytesIO
+
+import torch
+from diffusers import EulerDiscreteScheduler, StableDiffusionPipeline
+from fastapi import FastAPI, HTTPException
+from fastapi.responses import Response
+
+from ray import serve
+from ray.serve.handle import DeploymentHandle
+
+# Create a FastAPI instance to handle HTTP parsing and validation.
+# Learn more: https://docs.ray.io/en/latest/serve/http-guide.html#fastapi-http-deployments.
+app = FastAPI()
+
+# Set up the Ray Serve logger.
+logger = logging.getLogger("ray.serve")
+
+# A serve deployment contains business logic or an ML model to handle incoming requests.
+# It consists of a number of replicas that are individual copies of the class or function defined within it.
+@serve.deployment(num_replicas=1)
+@serve.ingress(app)
+class APIIngress:
+    def __init__(self, diffusion_model_handle) -> None:
+        self.handle = diffusion_model_handle
+
+    # Raise an error if a user sends a request to the root path.
+    @app.get("/")
+    async def generate(self):
+        raise HTTPException(
+            status_code=400, detail="'/' not supported, use '/imagine' instead."
+        )
+
+    # Handle API requests to the `/imagine` route.
+    @app.get(
+        "/imagine",
+        responses={200: {"content": {"image/png": {}}}},
+        response_class=Response,
+    )
+    async def generate(self, prompt: str, img_size: int = 512):
+        if not prompt:
+            raise HTTPException(status_code=400, detail="Missing 'prompt' parameter.")
+
+        image = await self.handle.generate.remote(prompt, img_size=img_size)
+
+        file_stream = BytesIO()
+        image.save(file_stream, "PNG")
+        return Response(content=file_stream.getvalue(), media_type="image/png")
+
+
+@serve.deployment(
+    ray_actor_options={"num_gpus": 1, 
+        "num_cpus": 1,  # Set the number of GPUs and CPUs required for each model replica.
+        "accelerator_type": "A10G"},   # Set accelerator type based on GPU type to use (T4, A10G, L4, V100, A100-40G or A100-80G)
+    max_concurrent_queries=2,   # Maximum number of queries that are sent to a replica of this deployment without receiving a response.
```

Comment:
how did you arrive at this? seems low

Reply:
max_concurrent_queries? Stable diffusion doesn't really do batching and there's not much asyncio. It should be 1 but we don't support that with auto-scaling until some 2.10 improvements land.